sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NormalizeFloatingNumbers.scala

```diff
@@ -116,6 +116,15 @@ object NormalizeFloatingNumbers extends Rule[LogicalPlan] {
     case CreateMap(children, useStringTypeWhenEmpty) =>
       CreateMap(children.map(normalize), useStringTypeWhenEmpty)
 
+    case If(cond, trueValue, falseValue) =>
```

Comment:
Although `cond` is `boolean` typed expression. I'm wondering what happens when we have a deeper expression at `cond`. Maybe, `normalize(cond)` is required here?

Reply:
Based on the doc of `NormalizeFloatingNumbers`, we handle that well during comparison. So the boolean condition is okay. We need normalize `trueValue` and `falseValue` because they could be passed through up to binary format in grouping/join/window partition.

Reply:
What happen when we have `If(If(If(cond, trueValue1, falseValue1), trueValue2, falseValue2), trueValue3, falseValue3)`? We don't need to invoke `normalize` at `trueValue1~2` and `falseValue1~2`?

Reply:
I think we only need to have normalization on `trueValue3` and `falseValue3`, as they are the only results passing through the top-level `If`. The inner `If`s are evaluated during executing the top-level `If`.

Reply:
Got it. I was confused a little. Thank you, @viirya .